File: part1/week_01_1/python/percolation.py
#!/usr/local/bin/python
# encoding: utf-8
from union_find import WeightedQuickFindUF
import random

import random


class Percolation(object):
    percolation_result = False

    def __init__(self, n):
        """create N-by-N grid, with all sites blocked"""
        self.n = n
        self.items_count = n * n
        self.uf = WeightedQuickFindUF(self.items_count + 2)
        self.top_index = self.items_count
        self.bottom_index = self.items_count + 1

        self.open_sites = [False] * (self.items_count + 2)
        self.open_sites[-1] = True
        self.open_sites[-2] = False

    def open_by_index(self, index):
        self.open(*self.index_to_ij(index))

    def open(self, i, j):
        index = self.ij_to_index(i, j)
        self.open_sites[index] = True
        self.check_nodes_around(i, j)
        if i == 0:
            self.uf.union(index, self.top_index)
        elif i == self.n - 1:
            self.uf.union(index, self.bottom_index)

    def check_nodes_around(self, i, j):
        item_index = self.ij_to_index(i, j)

        # bottom
        if i > 0 and self.is_open(i - 1, j):
            bottom_index = self.ij_to_index(i - 1, j)
            self.uf.union(item_index, bottom_index)
        # top
        if i < self.n - 1 and self.is_open(i + 1, j):
            top_index = self.ij_to_index(i + 1, j)
            self.uf.union(item_index, top_index)
        # left
        if j > 0 and self.is_open(i, j - 1):
            left_index = self.ij_to_index(i, j - 1)
            self.uf.union(item_index, left_index)
        # right
        if j < self.n - 1 and self.is_open(i, j + 1):
            right_index = self.ij_to_index(i, j + 1)
            self.uf.union(item_index, right_index)

    def is_open(self, i, j):
        """is site (row i, column j) open?"""
        return self.ij_to_index(i, j) in self.open_sites

    def is_full(self, i, j):
        """is site (row i, column j) full?"""
        return self.uf.connected(self.top_index, self.index(i, j))

    def percolates(self):
        """does the system percolate?"""
        return self.uf.connected(self.top_index, self.bottom_index)

    def ij_to_index(self, i, j):
        return self.n * i + j

    def index_to_ij(self, index):
        return index / self.n, index % self.n

    def visualize(self):
        return [['0' if self.is_open(i, j) else 'X' for j in xrange(self.n)] for i in xrange(self.n)]


# def ij_to_index(n, i, j):
#     return n * i + j

# def index_to_ij(n, index):
#     return index / n, index % n

def solution(n, t):
    p = Percolation(n)

    sites = random.sample(range(n * n), t)
    for site in sites:
        p.open_by_index(site)
        print 'opening site %d' % site
        if p.percolates():
            p.percolation_result = True
            print 'system percolates!'
            break

    return p


if __name__ == '__main__':
    for line in solution(20, 30).visualize():
        print line


class IndexOutOfBoundsException(Exception):
    pass


class IllegalArgumentException(Exception):
    pass

File: part1/week_01_1/java/Percolation.java
// import java.lang.IndexOutOfBoundsException;
// import java.lang.IllegalArgumentException;
// 15:50
public class Percolation {
    private int n;
    private int top_site;
    private int bottom_site;
    private boolean[][] sites;
    private WeightedQuickUnionUF uf;

    private void checkArgs(int i, int j) {
        if (i < 1 || i > n || j < 1 || j > n) {
            throw new IndexOutOfBoundsException();
        }
    }
    private void checkNodesAround(int i, int j) {
        int item_index = ijToIndex(i, j);
        if (i > 1 && isOpen(i - 1, j)) uf.union(item_index, ijToIndex(i - 1, j)); // bottom
        if (i < n && isOpen(i + 1, j)) uf.union(item_index, ijToIndex(i + 1, j)); // top
        if (j > 1 && isOpen(i, j - 1)) uf.union(item_index, ijToIndex(i, j - 1)); // left
        if (j < n && isOpen(i, j + 1)) uf.union(item_index, ijToIndex(i, j + 1)); // right
    }
    private int ijToIndex(int i, int j) {
        return n * (i - 1) + j;
    }

    public Percolation(int N) {
        // create N-by-N grid, with all sites blocked
        if (N <= 0) {
            throw new IllegalArgumentException();
        }

        int uf_size = N * N + 2;
        n = N;
        uf = new WeightedQuickUnionUF(uf_size);
        bottom_site = 0;
        top_site = N * N + 1;
        sites = new boolean[n][];

        for (int i = 0; i < n; i++) {
            sites[i] = new boolean[n];
            for (int j = 0; j < n; j++) {
                sites[i][j] = false;
            }
        }
    }
    public void open(int i, int j) {
        // open site (row i, column j) if it is not open already
        checkArgs(i, j);
        int index = ijToIndex(i, j);
        sites[i - 1][j - 1] = true;
        checkNodesAround(i, j);
        if (i == 1) uf.union(index, top_site);
        if (i == n) uf.union(index, bottom_site);

    }
    public boolean isOpen(int i, int j) {
        // is site (row i, column j) open?
        checkArgs(i, j);

        return sites[i - 1][j - 1];
    }
    public boolean isFull(int i, int j) {
        // is site (row i, column j) full?
        checkArgs(i, j);

        return uf.connected(top_site, ijToIndex(i, j));
    }
    public boolean percolates() {
        // does the system percolate?
        return uf.connected(top_site, bottom_site);
    }

    public static void main(String[] args) {
        // test client (optional)
        int i, j, k;
        int N = Integer.parseInt(args[0]);
        int sitesCount = N * N;
        int[] sitesToOpen = new int[sitesCount];

        for (k = 0; k < sitesCount; k++) {
            sitesToOpen[k] = k + 1;
        }
        StdRandom.shuffle(sitesToOpen);

        Percolation percolation = new Percolation(N);

        k = 0;
        while (k < sitesCount) {
            i = (sitesToOpen[k] - 1) / N;
            j = (sitesToOpen[k] - 1) % N;
            percolation.open(i + 1, j + 1);
            k++;
            if (percolation.percolates()) {
                break;
            }
        }
        StdOut.println("System percolates after " + k + " sites opened");
    }
}
